rest-api/offline/public_metrics_export.py

```diff
@@ -0,0 +1,204 @@
+         WHEN YEARS_OLD[:now, date_of_birth] <= 17 THEN '0-17'
+         WHEN YEARS_OLD[:now, date_of_birth] <= 25 THEN '18-25'
+         WHEN YEARS_OLD[:now, date_of_birth] <= 35 THEN '26-35'
+         WHEN YEARS_OLD[:now, date_of_birth] <= 45 THEN '36-45'
+         WHEN YEARS_OLD[:now, date_of_birth] <= 55 THEN '46-55'
+         WHEN YEARS_OLD[:now, date_of_birth] <= 65 THEN '56-65'
+         WHEN YEARS_OLD[:now, date_of_birth] <= 75 THEN '66-75'
+         WHEN YEARS_OLD[:now, date_of_birth] <= 85 THEN '76-85'
+         ELSE '86+'
+        END age_range,
+        COUNT(*)
+      FROM participant_summary
+      WHERE {filter_test_sql}
+      GROUP BY 1;
+      """.format(filter_test_sql=_FILTER_TEST_SQL),
+      None, None),
+  _SqlAggregation(
+      'physicalMeasurements',
+      """
+      SELECT physical_measurements_status, COUNT(*)
+      FROM participant_summary
+      WHERE {filter_test_sql}
+      GROUP BY 1;
+      """.format(filter_test_sql=_FILTER_TEST_SQL),
+      lambda v: PhysicalMeasurementsStatus.lookup_by_number(v).name,
+      None),
+  _SqlAggregation(
+      'biospecimenSamples',
+      """
+      SELECT
+        CASE
+         WHEN biospecimen_status IS NULL THEN 'UNSET'
+         WHEN biospecimen_status = :unset_status THEN 'UNSET'
+         WHEN biospecimen_status = :created_status THEN 'UNSET'
+         ELSE 'COLLECTED'
+        END, COUNT(*)
+      FROM participant_summary
+      WHERE {filter_test_sql}
+      GROUP BY 1;
+      """.format(filter_test_sql=_FILTER_TEST_SQL),
+      None,
+      params={
+        'unset_status': OrderStatus.UNSET.number,
+        'created_status': OrderStatus.CREATED.number
+      }),
+  # TODO(calbach): Add healthcare_access, medical_history, medications,
+  # family_health once available.
+  _questionnaire_metric('questionnaireOnOverallHealth', 'questionnaire_on_overall_health'),
+  # Personal habits is a newer naming for lifestyle
+  _questionnaire_metric('questionnaireOnPersonalHabits', 'questionnaire_on_lifestyle'),
+  # Sociodemographics is a newer naming for 'the basics'
+  _questionnaire_metric('questionnaireOnSociodemographics', 'questionnaire_on_the_basics'),
+]
+
+
+class PublicMetricsExport(object):
+  """Exports data from the database needed to generate public registration metrics."""
+
+  @staticmethod
+  def export():
+    # TODO(calbach): Write the output to a given target destination rather than
+    # returning it.
+    out = {}
+    # Using a session here should put all following SQL invocations into a
+    # non-locking read transaction per
+    # https://dev.mysql.com/doc/refman/5.7/en/innodb-consistent-read.html
+    test_hpo = HPODao().get_by_name(TEST_HPO_NAME)
+    with database_factory.make_server_cursor_database().session() as session:
+      for (key, sql, valuef, params) in _SQL_AGGREGATIONS:
+        sql = replace_years_old(sql)
+        out[key] = []
+        p = {
+          'now': clock.CLOCK.now(),
```

Comment:
Nit: store now() outside this loop and reuse it here, so we're using consistent timestamps across queries.

Reply:
Yikes, good catch.